// powerbi-visuals-tools/lib/CommandManager.js
import { createCertificate } from './CertificateTools.js';
import ConsoleWriter from './ConsoleWriter.js';
import VisualManager from './VisualManager.js';
export default class CommandManager {
    static async start(options, rootPath) {
        const webpackOptions = {
            devMode: true,
            devtool: "eval-source-map",
            generateResources: true,
            generatePbiviz: false,
            minifyJS: false,
            minify: false,
            devServerPort: options.port,
            stats: options.stats,
            skipApiCheck: options.skipApi,
            allLocales: options.allLocales,
            pbivizFile: options.pbivizFile,
        };
        const visualManager = new VisualManager(rootPath);
        await visualManager.prepareVisual(options.pbivizFile);
        await visualManager.validateVisual();
        await visualManager.initializeWebpack(webpackOptions);
        visualManager.startWebpackServer(options.drop);
    }
    static async lint(options, rootPath) {
        const visualManager = new VisualManager(rootPath);
        await visualManager.prepareVisual();
        await visualManager.runLintValidation(options);
    }
    static async package(options, rootPath) {
        if (!options.pbiviz && !options.resources) {
            ConsoleWriter.error('Nothing to build. Cannot use --no-pbiviz without --resources');
            process.exit(1);
        }
        const webpackOptions = {
            devMode: false,
            generateResources: options.resources,
            generatePbiviz: options.pbiviz,
            minifyJS: options.minify,
            minify: options.minify,
            compression: options.compression,
            stats: options.stats,
            skipApiCheck: options.skipApi,
            allLocales: options.allLocales,
            pbivizFile: options.pbivizFile,
            certificationAudit: options.certificationAudit,
            certificationFix: options.certificationFix,
        };
        const lintOptions = {
            verbose: options.verbose,
            fix: options.fix,
        };
        const visualManager = new VisualManager(rootPath);
        const visual = await visualManager.prepareVisual(options.pbivizFile);
        await visual.runLintValidation(lintOptions);
        await visual.validateVisual(options.verbose);
        await visual.initializeWebpack(webpackOptions)
            .then(manager => manager.generatePackage(options.verbose));
    }
    static new({ force, template }, name, rootPath) {
        const generateOptions = {
            force: force,
            template: template
        };
        VisualManager.createVisual(rootPath, name, generateOptions);
    }
    static async info(rootPath) {
        const visualManager = new VisualManager(rootPath);
        await visualManager.prepareVisual();
        await visualManager.displayInfo();
    }
    static async installCert() {
        await createCertificate();
    }
}
